crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -247,6 +286,184 @@ impl IndexerReader {
+                if descending_order {
+                    boxed_query =
+                        boxed_query.filter(checkpoints::sequence_number.lt(cursor as i64));
+                } else {
+                    boxed_query =
+                        boxed_query.filter(checkpoints::sequence_number.gt(cursor as i64));
+                }
+            }
+            if descending_order {
+                boxed_query = boxed_query.order_by(checkpoints::sequence_number.desc());
+            } else {
+                boxed_query = boxed_query.order_by(checkpoints::sequence_number.asc());
+            }
+
+            boxed_query
+                .limit(limit as i64)
+                .load::<StoredCheckpoint>(conn)
+        })
+    }
+
+    pub fn get_checkpoints(
+        &self,
+        cursor: Option<u64>,
+        limit: usize,
+        descending_order: bool,
+    ) -> Result<Vec<sui_json_rpc_types::Checkpoint>, IndexerError> {
+        self.get_checkpoints_from_db(cursor, limit, descending_order)?
+            .into_iter()
+            .map(sui_json_rpc_types::Checkpoint::try_from)
+            .collect()
+    }
+
+    pub fn get_transaction(
+        &self,
+        digest: TransactionDigest,
+    ) -> Result<Option<StoredTransaction>, IndexerError> {
+        self.run_query(|conn| {
+            transactions::table
+                .filter(transactions::transaction_digest.eq(digest.into_inner().to_vec()))
+                .first::<StoredTransaction>(conn)
+                .optional()
+        })
+    }
+
+    pub fn multi_get_transactions(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> Result<Vec<StoredTransaction>, IndexerError> {
+        let digests = digests
+            .iter()
+            .map(|digest| digest.inner().to_vec())
+            .collect::<Vec<_>>();
+        self.run_query(|conn| {
+            transactions::table
+                .filter(transactions::transaction_digest.eq_any(digests))
+                .load::<StoredTransaction>(conn)
+        })
+    }
+
+    pub fn multi_get_transaction_block_response(
+        &self,
+        digests: &[TransactionDigest],
+        options: &sui_json_rpc_types::SuiTransactionBlockResponseOptions,
+    ) -> Result<Vec<sui_json_rpc_types::SuiTransactionBlockResponse>, IndexerError> {
+        self.multi_get_transactions(digests)?
+            .into_iter()
+            .map(|transaction| transaction.try_into_sui_transaction_block_response(options, self))
+            .collect::<Result<Vec<_>, _>>()
+    }
+
+    pub async fn multi_get_transaction_block_response_async(
+        &self,
+        digests: Vec<TransactionDigest>,
+        options: sui_json_rpc_types::SuiTransactionBlockResponseOptions,
+    ) -> Result<Vec<sui_json_rpc_types::SuiTransactionBlockResponse>, IndexerError> {
+        self.spawn_blocking(move |this| {
+            this.multi_get_transaction_block_response(&digests, &options)
+        })
+        .await
+    }
+
+    pub fn get_transaction_events(
```

Comment:
should we make these non-spawn-blocking functions private so other modules won't accidentally call into it? Namely we always expose async only functions in indexer-reader

Reply:
That's a good idea...although it does potentially cause an issue with the GetModule trait impls and the like...we'd likely need to think about a good way to gate that functionality as well